File: ports/go-control-plane/ids.go
package controlplane

import (
	"encoding/hex"
	"errors"
	"fmt"
)

// ErrZeroID reports a raw identifier of zero, which no constructor issues: a
// decoded id of zero from a transport frame is a protocol error, not a handle.
var ErrZeroID = errors.New("identifier raw value 0 is never issued")

// AgentID identifies a logical agent registered in a control plane. Thousands
// may exist; none implies a thread or an OS process. The type exists so an
// AgentID can never be assigned to a TaskID slot by accident: they are distinct
// named types with no conversion path and private fields.
type AgentID struct{ raw uint64 }

// NewAgentID constructs an id from a counter value; zero is rejected.
func NewAgentID(raw uint64) (AgentID, error) {
	if raw == 0 {
		return AgentID{}, ErrZeroID
	}
	return AgentID{raw}, nil
}

// Raw exposes the counter value for placement keys and telemetry.
func (a AgentID) Raw() uint64 { return a.raw }

func (a AgentID) String() string { return fmt.Sprintf("agent-%d", a.raw) }

// TaskID identifies one unit of work admitted to a control plane.
type TaskID struct{ raw uint64 }

// NewTaskID constructs an id from a counter value; zero is rejected.
func NewTaskID(raw uint64) (TaskID, error) {
	if raw == 0 {
		return TaskID{}, ErrZeroID
	}
	return TaskID{raw}, nil
}

// Raw exposes the counter value; it is also the rendezvous placement key.
func (t TaskID) Raw() uint64 { return t.raw }

func (t TaskID) String() string { return fmt.Sprintf("task-%d", t.raw) }

// ShardID names one logical placement partition — an in-process queue lane,
// never a host.
type ShardID struct{ raw uint64 }

// NewShardID constructs an id from a partition index; zero is rejected so the
// zero value stays unusable, matching every other id here.
func NewShardID(raw uint64) (ShardID, error) {
	if raw == 0 {
		return ShardID{}, ErrZeroID
	}
	return ShardID{raw}, nil
}

// Raw exposes the zero-based partition index callers use to select a queue.
func (s ShardID) Raw() uint64 { return s.raw }

func (s ShardID) String() string { return fmt.Sprintf("shard-%d", s.raw) }

// LeaseEpoch is minted by a LeaseTable on every grant. Two handles with equal
// epochs can never both be live for one task because grants bump the epoch
// monotonically under the table's exclusivity rule. The constructor is
// unexported on purpose: only a LeaseTable may mint one.
type LeaseEpoch struct{ raw uint64 }

func newLeaseEpoch(raw uint64) LeaseEpoch { return LeaseEpoch{raw} }

// Raw exposes the counter value for audit logs and cross-checking receipts.
func (e LeaseEpoch) Raw() uint64 { return e.raw }

func (e LeaseEpoch) String() string { return fmt.Sprintf("epoch-%d", e.raw) }

// IdempotencyKey deduplicates submissions. Callers may supply one; otherwise
// Derive builds it from task identity so a retried submission is recognized as
// a duplicate while a genuinely new submission is not.
type IdempotencyKey [16]byte

// Derive deterministically mixes two 64-bit halves into a key, matching
// agent-fabric's IdempotencyKey::derive byte for byte.
func DeriveIdempotencyKey(a, b uint64) IdempotencyKey {
	var k IdempotencyKey
	bePutUint64(k[0:8], Mix64(a))
	bePutUint64(k[8:16], Mix64(b))
	return k
}

// Hex renders the full key; this exact string appears in parity vectors.
func (k IdempotencyKey) Hex() string { return hex.EncodeToString(k[:]) }

// Mix64 is the SplitMix64 finalizer shared by shard rendezvous scoring and key
// derivation. It lives in one place because those call sites need exactly this
// avalanche behaviour and nothing more, and parity requires that nothing
// substitutes a different mixer.
func Mix64(x uint64) uint64 {
	x += 0x9E3779B97F4A7C15
	x = (x ^ (x >> 30)) * 0xBF58476D1CE4E5B9
	x = (x ^ (x >> 27)) * 0x94D049BB133111EB
	return x ^ (x >> 31)
}

func bePutUint64(b []byte, v uint64) {
	for i := 7; i >= 0; i-- {
		b[i] = byte(v)
		v >>= 8
	}
}

func beUint64(b []byte) uint64 {
	var v uint64
	for _, c := range b {
		v = v<<8 | uint64(c)
	}
	return v
}

File: integrations/agent-fabric/src/ids.rs
//! Typed identifiers and the 64-bit mixing primitive shared by shard placement and key
//! derivation.
//!
//! Every identifier is a distinct type on purpose: an `AgentId` must never be assignable to a
//! `TaskId` slot by accident, so there are no `From` conversions between id types and the inner
//! integer is private. Zero is never issued, so `from_raw(0)` is `None` — a decoded id of zero
//! from a transport frame is a protocol error, not a valid handle.

use std::fmt;

macro_rules! plain_id {
    ($(#[$doc:meta])* $name:ident, $prefix:literal) => {
        $(#[$doc])*
        #[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
        pub struct $name(u64);

        impl $name {
            /// Constructs a typed id from a raw counter value. `0` is never issued.
            pub fn from_raw(raw: u64) -> Option<Self> {
                if raw == 0 { None } else { Some(Self(raw)) }
            }

            pub(crate) fn new(raw: u64) -> Self {
                Self(raw)
            }

            pub fn raw(self) -> u64 {
                self.0
            }
        }

        impl fmt::Display for $name {
            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
                write!(f, concat!($prefix, "-{}"), self.0)
            }
        }
    };
}

plain_id!(
    /// A logical agent registered in the fabric. Thousands may exist; none implies a thread or
    /// process.
    AgentId,
    "agent"
);
plain_id!(
    /// A unit of work admitted to the fabric.
    TaskId,
    "task"
);
plain_id!(
    /// A logical placement partition. Shards are in-process queues, not hosts.
    ShardId,
    "shard"
);

/// Epoch counter minted by the lease table on every grant or renewal. Two handles with equal
/// epochs can never both be live for one task because grants bump the epoch monotonically under
/// the table's exclusivity rule.
#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
pub struct LeaseEpoch(u64);

impl LeaseEpoch {
    pub(crate) fn new(raw: u64) -> Self {
        Self(raw)
    }

    pub fn raw(self) -> u64 {
        self.0
    }
}

impl fmt::Display for LeaseEpoch {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "epoch-{}", self.0)
    }
}

/// Deduplication key for submissions. Callers may supply one; otherwise the fabric derives it
/// from task identity and payload digest, so a retried *submission* (same key) is recognized as
/// a duplicate while a genuinely new submission is not.
#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
pub struct IdempotencyKey([u8; 16]);

impl IdempotencyKey {
    /// Deterministically derives a key by mixing two 64-bit halves. Used when a caller does not
    /// name a key explicitly.
    pub fn derive(a: u64, b: u64) -> Self {
        let mut k = [0u8; 16];
        k[..8].copy_from_slice(&mix64(a).to_be_bytes());
        k[8..].copy_from_slice(&mix64(b).to_be_bytes());
        Self(k)
    }

    pub fn from_bytes(bytes: [u8; 16]) -> Self {
        Self(bytes)
    }

    pub fn as_bytes(&self) -> &[u8; 16] {
        &self.0
    }

    pub fn hex(&self) -> String {
        let mut s = String::with_capacity(32);
        for b in self.0 {
            s.push_str(&format!("{b:02x}"));
        }
        s
    }
}

impl fmt::Display for IdempotencyKey {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        // Short display keeps logs readable; full fidelity lives in hex().
        let mut s = String::with_capacity(8);
        for b in self.0.iter().take(4) {
            s.push_str(&format!("{b:02x}"));
        }
        write!(f, "idem-{s}…")
    }
}

/// SplitMix64 finalizer. Shared here rather than in a rng module because shard rendezvous
/// scoring, key derivation and the simulator's PRNG all need exactly this avalanche behaviour
/// and nothing more.
pub fn mix64(mut x: u64) -> u64 {
    x = x.wrapping_add(0x9E37_79B9_7F4A_7C15);
    x = (x ^ (x >> 30)).wrapping_mul(0xBF58_476D_1CE4_E5B9);
    x = (x ^ (x >> 27)).wrapping_mul(0x94D0_49BB_1331_11EB);
    x ^ (x >> 31)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn zero_is_never_a_valid_raw_id() {
        assert!(AgentId::from_raw(0).is_none());
        assert!(AgentId::from_raw(1).is_some());
    }

    #[test]
    fn distinct_id_types_do_not_interconvert() {
        // The claim is compile-time: these are different nominal types with no From/Into path.
        let a = AgentId::from_raw(3).expect("nonzero");
        let t = TaskId::from_raw(3).expect("nonzero");
        assert_ne!(a.to_string(), t.to_string(), "display names the kind");
    }

    #[test]
    fn derived_keys_are_deterministic_and_collision_resistant_for_trivial_inputs() {
        assert_eq!(IdempotencyKey::derive(1, 2), IdempotencyKey::derive(1, 2));
        assert_ne!(IdempotencyKey::derive(1, 2), IdempotencyKey::derive(2, 1));
        assert_eq!(IdempotencyKey::derive(1, 2).as_bytes().len(), 16);
    }

    #[test]
    fn mix64_is_a_bijective_finalizer_on_sample_points() {
        // Not a proof of bijectivity, but catches sign/truncation slips: distinct inputs stay
        // distinct and the function is stable across calls.
        for x in [0u64, 1, u64::MAX, 0xDEAD_BEEF] {
            assert_eq!(mix64(x), mix64(x));
        }
        assert_ne!(mix64(0), mix64(1));
    }
}
